serialization.go

```diff
@@ -248,60 +248,104 @@ func IsPreIssuer(issuer *x509.Certificate) bool {
 	return false
 }
 
-// LogEntryFromLeaf converts a LeafEntry object (which has the raw leaf data after JSON parsing)
-// into a LogEntry object (which includes x509.Certificate objects, after TLS and ASN.1 parsing).
-// Note that this function may return a valid LogEntry object and a non-nil error value, when
-// the error indicates a non-fatal parsing error.
-func LogEntryFromLeaf(index int64, leafEntry *LeafEntry) (*LogEntry, error) {
-	var leaf MerkleTreeLeaf
-	if rest, err := tls.Unmarshal(leafEntry.LeafInput, &leaf); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal MerkleTreeLeaf for index %d: %v", index, err)
+// UnmarshaledLeafEntry represents LeafEntry after TLS unmarshaling, or before
+// TLS marshaling.
+type UnmarshaledLeafEntry struct {
+	// Leaf is a MerkleTreeLeaf corresponding to LeafEntry.LeafInput.
+	Leaf MerkleTreeLeaf
+	// Cert is either certificate or pre-certificate.
+	Cert ASN1Cert
+	// Chain is the issuing certificate chain, starting with the issuer of Cert.
+	Chain []ASN1Cert
+	// IsPrecert determines what the Cert field contains. It corresponds to
+	// Leaf.TimestampedEntry.EntryType.
+	IsPrecert bool
+}
+
+// UnmarshalLeafEntry TLS-unmarshals the passed in LeafEntry, and stores the
+// result in the passed in UnmarshaledLeafEntry.
+func UnmarshalLeafEntry(entry *LeafEntry, ret *UnmarshaledLeafEntry) error {
```

Comment:
This should return `(*UnmarshaledLeafEntry, error)`, instead of using an out parameter.

Reply:
Done.